src/docbuilder/rustwide_builder.rs

```diff
@@ -95,12 +96,43 @@ impl RustwideBuilder {
         // Ignore errors if detection fails.
         let old_version = self.detect_rustc_version().ok();
 
+        let mut targets_to_install = TARGETS
+            .iter()
+            .map(|t| t.to_string())
+            .collect::<HashSet<_>>();
+        let installed_targets = match self.toolchain.installed_targets(&self.workspace) {
+            Ok(targets) => targets,
+            Err(err) => {
+                if let Some(&ToolchainError::NotInstalled) = err.downcast_ref::<ToolchainError>() {
+                    Vec::new()
+                } else {
+                    return Err(err);
+                }
+            }
+        };
+
+        // The extra targets are intentionally removed *before* trying to update.
+        //
+        // If a target is installed locally and it goes missing the next update, rustup will block
+        // the update to avoid leaving the system in a broken state. This is not a behavior we want
+        // though when we also remove the target from the list managed by docs.rs: we want that
+        // target gone, and we don't care if it's missing in the next update.
+        //
+        // Removing it beforehand works fine, and prevents rustup from blocking the update later in
+        // the method.
+        for target in installed_targets {
+            if !targets_to_install.remove(&target) {
+                self.toolchain.remove_target(&self.workspace, &target)?;
+            }
+        }
+
         self.toolchain.install(&self.workspace)?;
-        for target in TARGETS {
+
+        for target in &targets_to_install {
```

Comment:
It looks like this will never update toolchains that are already installed, since they will be removed on line 124. I don't think this is the intended behavior.

Reply:
Those targets will be updated: once a target is installed rustup will keep it up to date even without running `rustup target add` every time. Before we were installing all the targets all the time because we didn't have the list of already installed targets, but now that we have it we can use that to cut the number of `rustup` calls we make.

Reply:
Ah, I misunderstood what call to rustup this was making; I thought it was calling `rustup toolchain install`, not `rustup target add`.